services/feedz/feedz.service.js

```diff
@@ -0,0 +1,148 @@
+'use strict'
+
+const Joi = require('joi')
+const { BaseJsonService, NotFound } = require('..')
+const {
+  renderVersionBadge,
+  searchServiceUrl,
+  stripBuildMetadata,
+  selectVersion,
+} = require('../nuget/nuget-helpers')
+
+const schema = Joi.object({
+  items: Joi.array()
+    .items(
+      Joi.object({
+        items: Joi.array().items(
+          Joi.object({
+            catalogEntry: Joi.object({
+              version: Joi.string().required(),
+            }).required(),
+          })
+        ),
+      }).required()
+    )
+    .max(1)
+    .default([]),
+}).required()
+
+const errorSchema = Joi.object({
+  message: Joi.string().required(),
+})
+
+const feedNotFoundMessage = /Feed .+ for .+ not found/
+
+class FeedzVersionService extends BaseJsonService {
+  static category = 'version'
+
+  static route = {
+    base: 'feedz',
+    pattern: ':which(v|vpre)/:organization/:repository/:packageName',
+  }
+
+  static examples = [
+    {
+      title: 'Feedz',
+      pattern: 'v/:organization/:repository/:packageName',
+      namedParams: {
+        organization: 'shieldstests',
+        repository: 'mongodb',
+        packageName: 'MongoDB.Driver.Core',
+      },
+      staticPreview: this.render({ version: '2.10.4' }),
+    },
+    {
+      title: 'Feedz (with prereleases)',
+      pattern: 'vpre/:organization/:repository/:packageName',
+      namedParams: {
+        organization: 'shieldstests',
+        repository: 'mongodb',
+        packageName: 'MongoDB.Driver.Core',
+      },
+      staticPreview: this.render({ version: '2.11.0-beta2' }),
+    },
+  ]
+
+  static defaultBadgeData = {
+    label: 'feedz',
+  }
+
+  static render(props) {
+    return renderVersionBadge(props)
+  }
+
+  apiUrl({ organization, repository }) {
+    return `https://f.feedz.io/${organization}/${repository}/nuget`
+  }
+
+  tryHandleError({ err }) {
+    // Feedz returns 404 Not Found both when organization/repository and package is not found:
+    //  a) If the org/repo is not found, it includes message in the response body,
+    //  b) If the package is not found or the path does not exist, it returns no body.
+    // The latter cases are indistinguishable so we assume that the package does not exist.
+
+    if (err.response && err.response.statusCode === 404) {
+      if (err.buffer === '') {
+        throw new NotFound({ prettyMessage: 'package not found' })
+      } else {
+        let json
+        try {
+          json = this._parseJson(err.buffer)
+          this.constructor._validate(json, errorSchema)
+        } catch {
+          throw err
+        }
+
+        if (feedNotFoundMessage.test(json.message)) {
+          throw new NotFound({ prettyMessage: 'repository not found' })
+        } else {
+          throw err
+        }
+      }
+    } else {
+      throw err
+    }
+  }
```

Comment:
Thanks for including inline comments on this one. Do you have a feel for how stable or brittle this logic/the error behavior may be?

I think it's nice to be able to provide very contextual error messages to users here, which in most cases will be helping them pinpoint which param value contains a typo. In most other cases though we've eschewed complex error handling logic like this and simply provide a `x or y not found` message.

The tradeoffs for us to consider here is whether the extra logic and behavior is worth maintaining to be able to provide separate `package not found` vs `repository not found` error messages vs. a simple `repository or package not found` message in 404 cases

cc @badges/shields-maintainers for other input

Reply:
Yeah, I'm also not really sure whether that is worth it (that's why I first added the tests and then modified everything - we can easily rollback). The gain is really small compared to the complication.

Regarding the stability - I am quite sure that this won't change in the coming future. Feedz hasn't really changed for like couple of years now and I don't think it will change. However, this is a Feedz extension and not something in NuGet Gallery API spec so this might be really just a maintenance burden with no real benefit.

Reply:
Yeah I don't think we should be looking for particular strings in responses, as that's quite fragile. We should just look for error codes as they're less likely to change, but not all APIs have unique codes per error. Failing that, we should probably just display a generic error message.

Reply:
Also in agreement with the above. It's good to have specific error messages, but not at the cost of significant added code complexity or parsing that's likely to break.